### ISSUE TITLE
Submodule to enable integration with Splunk

### ISSUE THREAD

**Issue by ocervell** [terraform-google-modules/terraform-google-log-export#17]
See list of subscriptions and list of logs from multiple projects

**Comment by iangelov** [comment]
resource "google_project_iam_custom_role" "td-log-consumer-role" {
  role_id     = "LogConsumerRole"
  title       = "Log Consumer Role"
  description = "A custom role to be used with the Splunk add-on for GCP"
  stage       = "ALPHA"

  permissions = [
    "pubsub.subscriptions.list",
    "resourcemanager.projects.get",
  ]
}

**Comment by aaron-lane** [comment]
@ocervell can you please clarify what the intent of this issue is?

**Comment by ocervell** [comment]
Yes, we've noticed when setting up Splunk that we can list PubSub subscriptions / projects, so the above code would enable this feature with automation.

**Comment by aaron-lane** [comment]
@ocervell if I understand correctly, this would be a new submodule which would create a service account and apply a custom IAM role to it, then output the email of the service account. Is that correct?

**Comment by xueshanf** [comment]
I may have run into a similar issue. Use the module as it is, I got `(gcloud.pubsub.subscriptions.pull) PERMISSION_DENIED: User not authorized to perform this action` when using the service account. If I grant more permissions (after generate key for the service account created by the module):

 ```
resource "google_project_iam_member" "svc_splunk_sink" {
  project = var.gcp_project_id
  role    = "roles/pubsub.subscriber"
  member  = "serviceAccount:${module.destination.pubsub_subscriber}"
}```

Then it works.

**Comment by xueshanf** [comment]
To make it work, I have to do this:
```
resource "google_project_iam_member" "svc_splunk_sink" {
  project = var.gcp_project_id
  role    = "roles/pubsub.subscriber"
  member  = "serviceAccount:${module.destination.pubsub_subscriber}"
}``` 

Use the module's service account and resource level roles, I got `(gcloud.pubsub.subscriptions.pull) PERMISSION_DENIED: User not authorized to perform this action`.

**Comment by morgante** [comment]
@xueshanf Is that even with the `create_subscriber` flag set to true?

@ocervell Can you confirm this example is sufficient for a functional Splunk config? https://github.com/terraform-google-modules/terraform-google-log-export/blob/master/examples/splunk-sink/main.tf

**Comment by xueshanf** [comment]
@ocervell yes the flag is true and I can validate on IAM console that the service account is created, and on the pub/sub console, the service account has the subscriber and viewer roles. According to Google’s pub/sub access control doc, this should work, but it seem not enough. The IAM role workaround I have is at project level which is not ideal.

**Comment by ocervell** [comment]
@morgante @aaron-lane I think we can just add it to the `splunk` example. We would need to grant the extra permissions to the subscriber service account created with the module.

@morgante last time I checked it worked with setting permissions on the pubsub topic level, but this would enable seeing other subscriptions (multiple log exports) if needed.

**Comment by morgante** [comment]
While we could add it at the project level, I'd prefer if someone did a bit of a deeper dive on *why* this isn't working before we overgrant permissions.

**Comment by ocervell** [comment]
I will check again if it works - no reason it shouldn't

**Comment by iangelov** [comment]
Hi chaps - given @ocervell and myself did some work on that, I thought it might be worth adding some additional context.

The goal of a module we developed at $work was to facilitate ingestion from Splunk (using the official GCP Pub/Sub add-on) while using a minimum set of privileges. What we ended up with allows the following:
- pubsub.subscriber on the Pub/Sub subscription which is bound to the consumer (Splunk) service account - used by Splunk to consume messages from Pub/Sub. This uses google_pubsub_subscription_iam_binding.
- pubsub.viewer on the Pub/Sub subscription which is bound to the consumer (Splunk) service account - used by Splunk to consume messages from Pub/Sub. This uses google_pubsub_subscription_iam_binding.
- pubsub.publisher on the Pub/Sub topic which is bound to the producer SA (that's the log export service account) - this is necessary so that the log export can publish messages to the Pub/Sub topic. This uses google_pubsub_topic_iam_binding.
- a custom role with only two permissions (pubsub.subscriptions.list and resourcemanager.projects.get), granted on the project level to the consumer (Splunk) service account - these are necessary so that Splunk can populate the mandatory drop-down menu selection choices. This uses google_project_iam_binding.

Code example:
```
resource "google_project_iam_custom_role" "security-logs-export-consumer-role" {
  role_id     = "SecurityLogsExportConsumerRole"
  title       = "Security Logs Export Consumer Role"
  description = "A custom role to be used with the Splunk add-on for GCP"
  stage       = "GA"

  permissions = [
    "pubsub.subscriptions.list",
    "resourcemanager.projects.get",
  ]
}
```

This is tested and working in an actual environment. Happy to answer any questions you might have, and provide additional code examples. In terms of where the role is created and where the service account is bound to it, this depends on the context, but I think it's not unreasonable to implement it at the project level, at least judging by observing how organisations approach setting up log exports.

**Comment by morgante** [comment]
Awesome, thanks for investigating! Do you think you'd be able to provide a PR with the required config for getting this working?

**Comment by Jberlinsky** [comment]
I've implemented this in https://github.com/terraform-google-modules/terraform-google-log-export/pull/53

**Comment by Jberlinsky** [comment]
Can this be closed, as #53  has been merged?
